// strapi-connector-firestore/lib/db/component-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentCollection = void 0;
class ComponentCollection {
    constructor(model) {
        this.model = model;
        this.collection = model.firestore.collection(model.collectionName);
    }
    throw() {
        throw new Error('Operations are not supported on component collections. ' +
            'This connector embeds components directly into the parent document.');
    }
    get converter() {
        return this.throw();
    }
    get path() {
        return this.throw();
    }
    autoId() {
        // This is used to generate IDs for components
        return this.collection.doc().id;
    }
    doc() {
        return this.throw();
    }
    get() {
        return this.throw();
    }
    where() {
        return this.throw();
    }
    orderBy() {
        return this.throw();
    }
    limit() {
        return this.throw();
    }
    offset() {
        return this.throw();
    }
}
exports.ComponentCollection = ComponentCollection;
